here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/query/WhereClauseBuilder.kt

```diff
@@ -160,6 +163,81 @@ class WhereClauseBuilder(private val request: ReadFeatures) {
         }
     }
 
+    private fun whereTags() {
+        val tagQuery = request.query.tags
+        if(tagQuery != null){
+            if (where.isNotEmpty()) {
+                where.append(" AND (")
+            } else {
+                where.append(" (")
+            }
+            whereNestedTags(tagQuery)
+            where.append(")")
+        }
+    }
+
+    private fun whereNestedTags(tagQuery: ITagQuery) {
+        when (tagQuery) {
+            is TagNot -> not(tagQuery.query, this::whereNestedTags)
+            is TagOr -> or(tagQuery.filterNotNull(), this::whereNestedTags)
+            is TagAnd -> and(tagQuery.filterNotNull(), this::whereNestedTags)
+            is TagQuery -> resolveSingleTagQuery(tagQuery)
+        }
+    }
+
+    private fun resolveSingleTagQuery(tagQuery: TagQuery) {
+        when (tagQuery) {
+            is TagExists -> {
+                where.append("$tagsAsJsonb ?? '${tagQuery.name}'")
```

Comment:
This is prone to SQL injections and ASCII zero injections. Values that are provided by the client, like tag-names, must only be used in prepared-statements! You can quote this alternatively, still, it may have a bad impact on performance, when it is executed multiple times, because it can't be cached by the query parser (very likely does not matter in this case, because tag queries will be slow anyway).